### ISSUE TITLE
Zeros in expected table: causing error for dot finder and other tools

### ISSUE THREAD

**Issue by agalitsyna** [open2c/cooltools#373]
**_Precedent:_** Expected table with zeros will cause NaNs in the dotfinder tables, which will cause an error with poor description and hard to debug. 

**_Proposed generalized solution:_** 
1. Add checks for presence of zeros in the expected table when parsing it in the tools; throw a warning
2. Add handling zeros in the expected table when it can allow the tool work on general

**_Potential inspiration for solving:_** 
Separating truly missing data and known missing data in saddles/compartments when aligning tracks with coolers: https://github.com/open2c/cooltools/pull/360

**Comment by agalitsyna** [comment]
Related report with the same issue in dot caller: https://github.com/open2c/cooltools/issues/364

**Comment by sergpolly** [comment]
@gibcus in the lab discovered a related issue independently - i'll try to get his data to reproduce the error and than fix it. Preliminary this might be caused by the tiny arms of acrocentric chromosomes that have very little to no data associated with them

**Comment by sergpolly** [comment]
addressed in dots with the latest commit https://github.com/open2c/cooltools/commit/d4518a1ec7e7cf25cf8a8ecc98d0a627cc2be90a

**Comment by gibcus** [comment]
Getting a different error now:
```
RemoteTraceback                           Traceback (most recent call last)
RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/home/gibcusj/miniconda3/envs/open2c-env/lib/python3.8/site-packages/pandas/core/indexing.py", line 1587, in _get_list_axis
    return self.obj._take_with_is_copy(key, axis=axis)
  File "/home/gibcusj/miniconda3/envs/open2c-env/lib/python3.8/site-packages/pandas/core/series.py", line 945, in _take_with_is_copy
    return self.take(indices=indices, axis=axis)
  File "/home/gibcusj/miniconda3/envs/open2c-env/lib/python3.8/site-packages/pandas/core/series.py", line 930, in take
    new_index = self.index.take(indices)
  File "/home/gibcusj/miniconda3/envs/open2c-env/lib/python3.8/site-packages/pandas/core/indexes/base.py", line 1189, in take
    taken = values.take(
  File "/home/gibcusj/miniconda3/envs/open2c-env/lib/python3.8/site-packages/pandas/core/arrays/interval.py", line 1081, in take
    left_take = take(
  File "/home/gibcusj/miniconda3/envs/open2c-env/lib/python3.8/site-packages/pandas/core/algorithms.py", line 1573, in take
    result = arr.take(indices, axis=axis)
IndexError: index -9223372036854775808 is out of bounds for axis 0 with size 23

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/gibcusj/miniconda3/envs/open2c-env/lib/python3.8/site-packages/multiprocess/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/home/gibcusj/miniconda3/envs/open2c-env/lib/python3.8/site-packages/multiprocess/pool.py", line 48, in mapstar
    return list(map(*args))
  File "/home/gibcusj/miniconda3/envs/open2c-env/lib/python3.8/site-packages/cooltools/api/dotfinder.py", line 1405, in <lambda>
    job = lambda tile: to_extract(to_score(tile))
  File "/home/gibcusj/miniconda3/envs/open2c-env/lib/python3.8/site-packages/cooltools/api/dotfinder.py", line 1037, in extract_scored_pixels
    threshold_of_pixels = threshold.iloc[lbin_idx]
  File "/home/gibcusj/miniconda3/envs/open2c-env/lib/python3.8/site-packages/pandas/core/indexing.py", line 1073, in __getitem__
    return self._getitem_axis(maybe_callable, axis=axis)
  File "/home/gibcusj/miniconda3/envs/open2c-env/lib/python3.8/site-packages/pandas/core/indexing.py", line 1616, in _getitem_axis
    return self._get_list_axis(key, axis=axis)
  File "/home/gibcusj/miniconda3/envs/open2c-env/lib/python3.8/site-packages/pandas/core/indexing.py", line 1590, in _get_list_axis
    raise IndexError("positional indexers are out-of-bounds") from err
IndexError: positional indexers are out-of-bounds
"""

The above exception was the direct cause of the following exception:

IndexError                                Traceback (most recent call last)
Input In [11], in <cell line: 9>()
      9 for cond in conditions:
     10     #### call dots using redefined kernels (without clustering) 
     11     print("calculating... " +cond)
---> 12     dots_round_df_all_5kb[cond] = cooltools.dots(
     13         clr_5kb[cond],
     14         expected=expected_5kb[cond],
     15         view_df=hg38_arms,
     16         kernels=kernels_round_5kb, # provide custom kernels
     17         max_loci_separation=20_000_000,
     18         clustering_radius=None,
     19         tile_size=5_000_000,
     20         nproc=32,
     21         lambda_bin_fdr=0.15,
     22         max_nans_tolerated=2
     23     )
     25     #### call dots using redefined kernels (with clustering, using the "galGal6 settings" on hg38) 
     26     print("calculating... " +cond)

File ~/miniconda3/envs/open2c-env/lib/python3.8/site-packages/cooltools/api/dotfinder.py:1630, in dots(clr, expected, expected_value_col, clr_weight_name, view_df, kernels, max_loci_separation, max_nans_tolerated, n_lambda_bins, lambda_bin_fdr, clustering_radius, cluster_filtering, tile_size, nproc)
   1628 # 3. Filter using FDR thresholds calculated in the histogramming step
   1629 time_start = time.perf_counter()
-> 1630 filtered_pixels = scoring_and_extraction_step(
   1631     clr,
   1632     expected,
   1633     expected_value_col=expected_value_col,
   1634     clr_weight_name=clr_weight_name,
   1635     tiles=tiles,
   1636     kernels=kernels,
   1637     ledges=ledges,
   1638     thresholds=threshold_df,
   1639     max_nans_tolerated=max_nans_tolerated,
   1640     loci_separation_bins=loci_separation_bins,
   1641     nproc=nproc,
   1642     bin1_id_name="bin1_id",
   1643     bin2_id_name="bin2_id",
   1644 )
   1645 elapsed_time = time.perf_counter() - time_start
   1646 logging.info(f"Done extracting enriched pixels in {elapsed_time:.3f} sec ...")

File ~/miniconda3/envs/open2c-env/lib/python3.8/site-packages/cooltools/api/dotfinder.py:1421, in scoring_and_extraction_step(clr, expected_indexed, expected_value_col, clr_weight_name, tiles, kernels, ledges, thresholds, max_nans_tolerated, loci_separation_bins, nproc, bin1_id_name, bin2_id_name)
   1417 try:
   1418     # consider using
   1419     # https://github.com/mirnylab/cooler/blob/9e72ee202b0ac6f9d93fd2444d6f94c524962769/cooler/tools.py#L59
   1420     filtered_pix_chunks = map_(job, tiles, **map_kwargs)
-> 1421     significant_pixels = pd.concat(filtered_pix_chunks, ignore_index=True)
   1422 finally:
   1423     if nproc > 1:

File ~/miniconda3/envs/open2c-env/lib/python3.8/site-packages/pandas/util/_decorators.py:331, in deprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper(*args, **kwargs)
    325 if len(args) > num_allow_args:
    326     warnings.warn(
    327         msg.format(arguments=_format_argument_list(allow_args)),
    328         FutureWarning,
    329         stacklevel=find_stack_level(),
    330     )
--> 331 return func(*args, **kwargs)

File ~/miniconda3/envs/open2c-env/lib/python3.8/site-packages/pandas/core/reshape/concat.py:368, in concat(objs, axis, join, ignore_index, keys, levels, names, verify_integrity, sort, copy)
    146 @deprecate_nonkeyword_arguments(version=None, allowed_args=["objs"])
    147 def concat(
    148     objs: Iterable[NDFrame] | Mapping[HashableT, NDFrame],
   (...)
    157     copy: bool = True,
    158 ) -> DataFrame | Series:
    159     """
    160     Concatenate pandas objects along a particular axis.
    161 
   (...)
    366     1   3   4
    367     """
--> 368     op = _Concatenator(
    369         objs,
    370         axis=axis,
    371         ignore_index=ignore_index,
    372         join=join,
    373         keys=keys,
    374         levels=levels,
    375         names=names,
    376         verify_integrity=verify_integrity,
    377         copy=copy,
    378         sort=sort,
    379     )
    381     return op.get_result()

File ~/miniconda3/envs/open2c-env/lib/python3.8/site-packages/pandas/core/reshape/concat.py:422, in _Concatenator.__init__(self, objs, axis, join, keys, levels, names, ignore_index, verify_integrity, copy, sort)
    420     objs = [objs[k] for k in keys]
    421 else:
--> 422     objs = list(objs)
    424 if len(objs) == 0:
    425     raise ValueError("No objects to concatenate")

File ~/miniconda3/envs/open2c-env/lib/python3.8/site-packages/multiprocess/pool.py:420, in <genexpr>(.0)
    412 result = IMapIterator(self)
    413 self._taskqueue.put(
    414     (
    415         self._guarded_task_generation(result._job,
   (...)
    418         result._set_length
    419     ))
--> 420 return (item for chunk in result for item in chunk)

File ~/miniconda3/envs/open2c-env/lib/python3.8/site-packages/multiprocess/pool.py:868, in IMapIterator.next(self, timeout)
    866 if success:
    867     return value
--> 868 raise value

IndexError: positional indexers are out-of-bounds
```

I used` pip install cooltools --upgrade`. Perhaps that did not get me the latest commit?

**Comment by Phlya** [comment]
@gibcus try `pip install --upgrade --force-reinstall --no-deps --no-cache-dir https://github.com/open2c/cooltools/archive/refs/heads/master.zip`

**Comment by gibcus** [comment]
Thanks @Phlya! That got me to a newer `cooltools-0.5.2`. 
Let's see if it runs!

**Comment by sergpolly** [comment]
That's pretty much the same error - so yes - you haven't updated the first time. Regarding versions - they're assigned to properly released versions. What I did - was just changed the master - on top of released 0.5.2 , but haven't released anything properly... So 0.5.2 by itself doesn't mean much in this case

**Comment by gibcus** [comment]
0.5.2 works!
________________________________
